microsoft/MixedRealityToolkit-Unity
# First set of active profile change fixes

This change is the first in a series of changes that will addresses the issues encountered when changing the active mixed reality toolkit configuration profile while the application is running.

The actual number of future PRs has not yet been determined. Test collateral and documentation will be included in the series as fixes enable them.

- Failure to recreate the UIRaycastCamera

~~Change from Destroy to DestroyImmediate as the service locator expects the UIRaycastCamera to me immediately removed in order to correctly re-create it.~~

- Handle data provider collection change during {Late]Update

~~This change handles the InvalidOperationException that is thrown when a collection is changed during enumeration, This prevents the code from incorrectly aborting.~~

- Handle service collection change during ExecuteOnAllServices<In|Reverse>Order

~~This change replaces for with foreach and handles the InvalidOperationException that is thrown the services collection is changed during enumeration. This prevents the code from incorrectly aborting.~~

In this PR, the behavior of runtime profile change is modified to first take the request, then destroy current profile services after all LateUpdate()s of current profile services have been called, and finally initialize the new services associated with the new profile in the next frame before any Update() for a service is called. This solves the problem of potentially performing the profile change in the middle of an Update or LateUpdate of a service which causes a variety of problems. Delaying the initialization of new services to the next frame also helps to resolve the problems related to Destroy().

Testing

- In editor using simulated hands
- Toggle between two profiles on a click event
- Confirm that interactable objects continue to function after re-acquiring the hand.
- Attempt to use UnityUI
  - Fails with far interaction
  - Works with near interaction
